// commitizen/dist/commitizen.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.staging = exports.init = exports.configLoader = exports.commit = exports.cache = exports.adapter = undefined;

var _adapter = require('./commitizen/adapter');

var adapter = _interopRequireWildcard(_adapter);

var _cache = require('./commitizen/cache');

var cache = _interopRequireWildcard(_cache);

var _commit = require('./commitizen/commit');

var _commit2 = _interopRequireDefault(_commit);

var _configLoader = require('./commitizen/configLoader');

var configLoader = _interopRequireWildcard(_configLoader);

var _init = require('./commitizen/init');

var _init2 = _interopRequireDefault(_init);

var _staging = require('./commitizen/staging');

var staging = _interopRequireWildcard(_staging);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) newObj[key] = obj[key]; } } newObj.default = obj; return newObj; } }

exports.adapter = adapter;
exports.cache = cache;
exports.commit = _commit2.default;
exports.configLoader = configLoader;
exports.init = _init2.default;
exports.staging = staging;
